### ISSUE TITLE
Activating grains based on work queue

### ISSUE THREAD

**Issue by Eldar1205** [dotnet/orleans#1410]
Before learning about Orleans I had the following design for an Azure cloud service that AFAIK Orleans runtime doesn't handle for me and I suggest to add that build in support:
1. My frontend is a web role en-queues work items to an Azure storage queue
2. My backend is a worker role that its instances are getting work items from the queue in configurable batches (1-32 mostly, due to Azure storage queue limits), invisibility time for the work item is set in the queue for "enough time" for processing to complete
3. Each work item is processed by a processing object isolated from the others
4. After the work item processing is complete the processing object en-queues another work item and then it's deleted from the Azure storage queue. If work item processing failed, it's remained in the queue so it's de-queued later again once invisibility time is over.

The motivation for a persistent reliable work queue is:
1. Have an "at least once" work item delivering guarantee, which is "exactly once" if there are no errors, and isn't processed in parallel by two processing objects as long as invisibility time is set for a work item such that processing would be finished before the work item is visible again.
2. Azure has auto-scaling option based on queue capacity, which is very useful to automatically scale-out work items processing based on their amount. Orleans currently doesn't come with an auto-scaling option, and in particular no auto-scaling based on grains workload. 

Here is how I imagine Orleans build-in support for such a scenario:
1. Introduce a mechanism to send grains commands using a persistent reliable storage, similar in semantics to Azure storage queue, let's call it "the work queue".
2. Frontend clients and backend grains should be able to en-queue work items to the work queue, in a "fire and forget" manner - the en-queue task should complete successfully if the work item was persisted to the work queue, without waiting for receiving grain to handle work item and complete it.
3. Orleans runtime will have agents pulling work items from the queue and dispatching work items to grains. This may activate the grains, like reminders may activate grains.
4. Grains that can accept work items should implement a dedicated interface, like the one we have for grains responding to reminders. The interface will have a method to accept a strongly-typed (e.g. generics) work item, represented as IWorkItem of TBody where TBody is metadata about the work item submitted by the entity that en-queued the work item, and the IWorkItem<> object itself also has data about en-queue time, de-queue count & other Azure queue useful properties + method to update TBody (Azure storage queue supports updating a message for state tracking).
5. Orleans runtime should set invisibility time for work items based on the grain execution timeout, so the grain can process the work item knowing it remains invisible in the queue as long as the grain has control and he isn't timing out. The grain should be told by Orleans he's timing out, and allow him to extend his time (like he can postpone deactivation) which in turn will also increase invisibility time on the work item.
6. Once the grain method that handles the work item finishes execution without errors, Orleans runtime should assume the work item was processed and therefore could delete its message from the queue. If execution finishes with error, the work item should remain in the queue until invisibility time is over and the work item is available for pulling agents.
7. If a grain en-queues a work item that will ultimately be dispatched to itself, it shouldn't cause dead locks like when a grain uses recursion. This is because the grain method that accepts work items has the single execution semantics, which means the pulling agent can sent the work item to the grain but it'll wait until he finishes current work item, unless [Reentrant] is used.
8. The pulling agents that dispatch work items to grains should use Orleans grain request existing infrastructure, to benefit from built in grains distribution and elasticity. If the dispatch fails or the accepting grain throws exception while processing the work item, Orleans marks the task on the pulling agent side as broken and the pulling agent can use that info to decide whether to delete the work item from the queue or let it become visible again. 
9. The pulling agents shouldn't "await" every single work item, but fan-out the batch of work items they got from the work queue and react to the success/failure of each one of them independently. 

Hope I've made myself clear, let's discuss this further and talk about alternative solutions and how to approach this using current Orleans features. I'm new to github issues so advises/reading material/tips for that are also welcome. Thanks for reading this far!

**Comment by centur** [comment]
:) This scenario is pretty common in Azure and Orleans Streams is what you need. Use `Orleans.Providers.Streams.AzureQueue.AzureQueueStreamProvider` (based on Azure Queues) and just send messages to stream\namespace combination - runtime will pull it and invoke grain method on listener. 

You may need to introduce intermediate "listener" grains which will receive a message and dispatch it to a specific interfaces you want in 4.

Other parts you described are mostly implementation details, but I think you can build it using just Orleans + Right stream provider + some custom logic you implements on grains. 

PS: I'm not convinced that doing some of these requirements (like fan-out on messages) are the right way to do such things (e.g. what would you do if out of 30 fan-out 20 will succeed and 10 will fail - you need to selectively mark messages as done\fail) but it's your call. We are doing very similar processing just using streams, but we have stricter requirements to messages ordering.

**Comment by Eldar1205** [comment]
If my grains accepts work items from the queue stream provider, and it
fails to process, will the work item be re-accepted by same grain in
same/another activation? By reading the AzureQueueStreamProvider code I
conclude that's not the case, it uses a "NoOpStreamDeliveryFailureHandler"

2016-02-08 0:36 GMT+02:00 Alexey Shcherbak <notifications@github.com>:

**Comment by Eldar1205** [comment]
Additionally, using a stream for dispatching work items is something I discussed with Sergey Bykov and we agreed this would be an anemic stream. The work items should be dispatched immediately as they are taken of the queue, and not cached/buffered. Failure/Success on the subscribing grain should affect the status of the message in the queue, and setting it's invisibility time so it wouldn't become visible while the grain is still processing is also something the stream provider don't handle. In short, the use case for streams appears to be transferring data, not meta data, though it may be possible by implementing a new stream provider based on a custom queue adapter. What do you think?

**Comment by centur** [comment]
Yes, valid points - streams are delivery mechanisms, they are not supposed to participate in your code logic flow. If your grain accepts work item - IMO it's now the grain responsibility to handle the error. I think, that it might be not a good idea to return SAME message to the queue, if processing fails. I'd rather re-queue different message (with same data if needed) to signal about prev message failure and trigger "recovery procedure". 

From my experience we had more problems with re-using messages and underlying mechanisms like making message "invisible" for any complex processing. It works ok for simple messaging and chaining but doesn't work well for any complex processing. So I would not rely on message invisibility of a specific implementation as a proper way to build my system. But keep in mind, we also have different requirements - we need partial message ordering within Id\namespace pair (and AzureQueue is not the best fit for us too - to guarantee that we will implement our own provider, table or streamstone-based).

And again (we had that discussion on Gitter with other guys) - this adds a leaky abstraction (and our implementation does that too tbh) :

You're building your system based on assumption that this given implementation works in a certain way. This means you can't easily swap implementations and you're in the land of fragility - because if current implementation changes it's behaviour - you never know how and when it will affect your own code. 
We were hit by similar issue when we switched from AzureQeueu to SimpleStream provider for local runs and tests - code was "working" but simpleStream provider introduced grain deadlocks.

**Comment by Eldar1205** [comment]
I agree with most of you said, except for not relying on invisibility. If
my processing fails and I requeue the work item, the requeue operation may
fail and I get at most once guarantee, so I will need to intoduce another
mechanism to ensure work items aren't lost, when Azure already provided me
with invisibilty that enable an at least once guarantee which is superior
to at most once because Orleans makes sure the same grain will handle both
work item duplicates sequentially preventing concurrency issues.
בתאריך 8 בפבר' 2016 02:41,‏ "Alexey Shcherbak" <notifications@github.com>
כתב:

**Comment by centur** [comment]
Any operation can fail. It's up to you what you consider more reliable. Again - if you'll bring such logic to a Stream Implementation - you will break the Liskov principle - you no longer will be able to replace stream provider because your implementation depends on a specific behaviour. Up to you - you can implement such provider using extensibility model as docs says, I just not sure how can you signal back to the provider that you handled the message so provider can delete it from queue. And this may fail to. You may have poisonous message that will always fail - and you need to count number of de-queueing. And also - you can have long running (longer than invisibility period) or infi-looped processing operation which will never inform stream provider about your success or fail with processing. Then you need to introduce circuit breaker pattern and you still need to rely on your own code to handle these things reliably. Sorry for opening this can of worms.

I'd prefer to rely on that my very simple code of saving\re-queueing message will succeed (with X number of retries, if needed) rather than taking such hard dependency on a specific provider implementation. You can take dependency on a given implementation if you're pulling messages from a queue in your code directly, but not when you're trying to use "want-to-be-unified" stream providers interface.

And on the delivery guarantees - as http://dotnet.github.io/orleans/Orleans-Streams/ says - Simple stream provider is at most once (event order is guaranteed), Azure Queue is at-least-once (order doesn't guaranteed).

**Comment by jason-bragg** [comment]
I suggest splitting out the queue processing aspects of the persistent stream provider into a public queue processing framework, then retrofitting the persistent stream provider to use this framework, while @Eldar1205 uses it to implement the capabilities described in this thread.

A necessary behavior in stream providers that use persistent queues (our PersistentStreamProvider for instance) is the need to read and process messages from a queue, i.e. queue processing.  This means that we need to perform queue processing in any case, so what we're really talking about, IMO, is how we implement queue processing.

Until now, queue processing has been an internal implementation detail of the PersistentStreamProvider, but if we introduce queue processing as a publicly supported set of Orleans capabilities, architected to use some sort of QueueProviders that act as integration points for different types of queues, we'd be able to address @Eldar1205's needs without necessarily requiring that a stream processing pattern be applied.

Once a queue processing framework is in place, the PersistentStreamProvider could be built on top of this queue processing, to leverage the work and reduce code duplication.  Additionally, other stream processing patterns (possibly violating Liskov's principle, but of utility none the less) could also be built on top of the queue processing framework with less code duplication.  As it stands now, any alternative stream providers that uses a persistent queue would need to duplicate, in its totality, the persistent stream provider's queue processing logic.

>you will break the Liskov principle 

The "Liskov principle" problem in streams has been there for a while and we've not found a good solution, so I'm going to ignore it for now, as I fear addressing it hear would essentially highjack this thread. :/

**Comment by gabikliot** [comment]
Regardless of the question whether streams are the right abstractions for your use case, we should fix what I consider as a bug in AzureQueueStreamProvider failing to guarantee at least once delivery.
Open a bug for that in #1416.

**Comment by gabikliot** [comment]
Bug #1416  was fixed via #1491.

**Comment by gabikliot** [comment]
This can be now closed, right @Eldar1205 ? Are you unblocked now?

**Comment by Eldar1205** [comment]
Yes, thank you
בתאריך 26 בפבר' 2016 9:36,‏ "Gabriel Kliot" <notifications@github.com> כתב:
